File: threeSum.cpp
#include <iostream>
#include <vector>
#include <set>
#include <algorithm>

using namespace std;

// https://leetcode.com/problems/3sum/description
class Solution {
public:
    int biSearch(vector<int>& v, int start, int value) {
        int low = start;
        int high = v.size() - 1;
        int mid;
        while (low <= high) {
            mid = (low + high) / 2;
            if (v[mid] == value) return mid;
            if (v[mid] > value) {
                high = mid - 1;
            } else low = mid + 1;
        }
        return -1;
    }

    vector<vector<int> > threeSum(vector<int>& vv) {
        vector<vector<int> > res;
        sort(vv.begin(), vv.end());
        vector<int> v;
        int count = 0;
        for (int i = 0; i < vv.size(); ++i) {
            if (count == 3 && vv[i] == vv[i-1]) continue;
            v.push_back(vv[i]);
            if (i > 0 && vv[i] != vv[i-1]) count = 0;
            ++count;
        }

        if (v.size() < 3) return res;
        int max = v[v.size() - 1];
        set<long> mark;

        for (int i = 0; i < v.size() - 1; ++i) {
            for (int j = i + 1; j < v.size(); ++j) {
                int itr = biSearch(v, j + 1, 0-v[i]-v[j]);
                if (itr != -1) {
                    long key = v[i] * max + v[j];
                    if (mark.find(key) == mark.end()) {
                        res.push_back(vector<int>{v[i], v[j], v[itr]});
                        mark.insert(key);
                    }
                }
            }
        }
        return res;
    }
};

int main(void) {
    vector<int> v{-1, 0, 1, 2, -1, -4};
    Solution s;
    vector<vector<int> > vvv = s.threeSum(v);
    for (int i = 0; i < vvv.size(); ++i) {
        for (int j = 0; j < vvv[i].size(); ++j)
            cout << vvv[i][j] << " ";
        cout << endl;
    }
    return 0;
}

File: go/threeSum.go
package main

import (
	"fmt"
	"sort"
)

// https://leetcode-cn.com/problems/3sum/

func threeSum0(nums []int) [][]int {
	var res [][]int
	sort.Ints(nums)
	for i, prei := 0, -1; i < len(nums)-2 && nums[i] <= 0; i++ {
		if prei >= 0 && nums[prei] == nums[i] {
			continue
		}
		for j, prej := i+1, -1; j < len(nums)-1 && nums[j] <= -nums[i]; j++ {
			if prej > i && nums[prej] == nums[j] {
				continue
			}
			if nums[i]+2*nums[j] > 0 {
				break
			}
			k := biSearch(-nums[i]-nums[j], nums[j+1:])
			if k >= 0 {
				res = append(res, []int{nums[i], nums[j], nums[k+j+1]})
			}
			prej = j
		}
		prei = i
	}
	return res
}

func threeSum(nums []int) [][]int {
	var res [][]int
	if len(nums) < 3 {
		return res
	}
	sort.Ints(nums)

	for i := 0; i < len(nums)-2 && nums[i] <= 0; i++ {
		if i != 0 && nums[i] == nums[i-1] {
			continue
		}
		for l, r := i+1, len(nums)-1; l < r; {
			for ; l < r && l > i+1 && nums[l] == nums[l-1]; l++ {
			}
			if l >= r {
				break
			}
			if nums[i]+nums[l]+nums[r] == 0 {
				res = append(res, []int{nums[i], nums[l], nums[r]})
				l++
				r--
			} else if nums[i]+nums[l]+nums[r] < 0 {
				l++
			} else {
				r--
			}
		}
	}
	return res
}

func biSearch(v int, nums []int) int {
	l, h := 0, len(nums)-1
	mid := 0
	for l <= h {
		mid = (l + h) / 2
		if nums[mid] == v {
			return mid
		}
		if nums[mid] > v {
			h = mid - 1
		} else {
			l = mid + 1
		}
	}
	return -1
}

func main() {
	cases := [][]int{
		{-1, 0, 1, 2, -1, -4},
		{-1, 0, 1, 2, -1, -4, -2, -3, 3, 0, 4},
		{0, 0, 0},
		{-2, 0, 0, 2, 2},
	}
	realCase := cases[0:]
	for i, c := range realCase {
		fmt.Println("## case", i)
		fmt.Println("I:", c)
		fmt.Println("O:", threeSum(c))
	}
}
